src/UCTNode.cpp

```diff
@@ -300,29 +301,43 @@ UCTNode* UCTNode::uct_select_child(int color) {
     auto parentvisits = size_t{0};
     for (const auto& child : m_children) {
         if (child->valid()) {
-            parentvisits += child->get_visits();
-            if (child->get_visits() > 0) {
-                total_visited_policy += child->get_score();
+            auto visits = child->get_visits();
+            auto psa = child->get_score();
+            if (visits > 0) {
+                parentvisits += visits;
+                total_visited_policy += psa;
+            }
+            else {
+                if (psa > best_unvisited_psa) {
+                    best_unvisited_psa = psa;
+                    best = child.get();
+                }
             }
         }
     }
 
     auto numerator = std::sqrt((double)parentvisits);
-    auto fpu_reduction = cfg_fpu_reduction * std::sqrt(total_visited_policy);
-    // Estimated eval for unknown nodes = original parent NN eval - reduction
-    auto fpu_eval = get_net_eval(color) - fpu_reduction;
+
+    if (best_unvisited_psa > -1000.0) {
+        // Estimated eval for unknown nodes = original parent NN eval - reduction
+        auto fpu_reduction = cfg_fpu_reduction * std::sqrt(total_visited_policy);
+        auto fpu_eval = get_net_eval(color) - fpu_reduction;
+        auto best_unvisited_puct = cfg_puct * best_unvisited_psa * numerator;
+        best_value = fpu_eval + best_unvisited_puct;
+    }
 
     for (const auto& child : m_children) {
         if (!child->valid()) {
             continue;
         }
-
-        float winrate = fpu_eval;
-        if (child->get_visits() > 0) {
-            winrate = child->get_eval(color);
+        auto visits = child->get_visits();
+        if (visits == 0) {
```

Comment:
caching visits isn't a speed up. this would be more cleanly written as

if (child->first_visit()) {
   continue;
}

auto denom = 1.0 + child->get_visits();